draft-aegis-mimi-arch-latest.md

```diff
@@ -560,17 +564,38 @@ Output:
 
 The above request has the following effect:
 
-* An arbitrary key pool with attributes matching the provided `appId`,
+// TODO: Should we really have isFinalKey or is that potentially a security
+problem and we should say that the pool should just return an error in the event
+there are no more keys.
+
+* A key queue with attributes matching the provided `appId`,
   `clientId`, `protocolVersion`, `cipherSuite` and `appLabel` is chosen. If
-  there is no such key pool, the request has no effect.
-* If the above key pool contains (non-final) key packages, a key package from
-  the pool is chosen (according to gatewey's policy), returned and deleted. The
-  `isFinalKey` flag in the output is set to false. The returned key package MUST
-  NOT be returned by any future fetch key package request.
-* Else, if there is a final key stored for the key pool, the output contains
+  there is no such key queue, the request has no effect.
+* If a non-final key in the queue is assigned to `receiver_client_id`, and that
+  key package is not expired then that key should be returned.
+* Else, if the above key queue contains (non-final) key packages, the first
+  non-expired key package that has not been assigned a receiver client id
+  should be assigned to the provided `receiver_client_id` and returned with the
+  `isFinalKey` flag set to false. Key packages SHOULD be sorted by the
+  `not-after` property of their `Lifetime` when choosing a key to return.
+* Else, if there is a final key stored for the key queue, the output contains
   this final key and the `isFinalKey` flag is set to true.
 * Else, the request has no effect.
 
+### Advancing a Key Package Queue 
+
```

Comment:
Maybe mention what "advancing" means. Like: Advancing past a key package is done by deleting the key package from the queue.

Reply:
I figured that was inferred, but can make that clearer